### PR TITLE
Blog: functional programming constructs

### PR DESCRIPTION
**Pull request by divyanshu013**
Feel free to leave any feedback. Check the [CI deploy link](https://deploy-preview-7--divyanshu-blog.netlify.com/blog/functional-programming-constructs/).

@sagarkarira @metagrover @dhruvdutt

### CODE REVIEW

**Review by dhruvdutt — APPROVED** [review]

**Review by dhruvdutt — APPROVED** [review]
Ship it. :shipit:

**Comment by divyanshu013** [comment]
TIL: you can't request people for review in personal repos unless they're collaborators.

File: content/blog/functional-programming-constructs/index.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+Functional programming avoids mutation of state and side effects. These may include modifying variables outside the scope of function, database or network operations, DOM manipulation and even using random numbers.
+
+#### Mutation
+
+Functional programming favors **value immutability**. Whenever we need to change the state in a program, we should create and track a new value for the state (instead of mutating the existing one).
+
+#### Side effects
+
+Its impossible to do anything useful in a software without doing any side effects, the existence or non existence of such a software would be equivalent. Instead, functional programming encourages us to make such side effects **more obvious** and **isolated**.
```
Comment (metagrover) [inline content/blog/functional-programming-constructs/index.md]:
```suggestion
It is impossible to do anything useful in a software without causing any side effects, the existence or non existence of such a software would be equivalent. Instead, functional programming encourages us to make such side effects **more obvious** and **isolated**.
```

File: content/blog/functional-programming-constructs/index.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+One of the ways of achieving this is by **favoring pure functions over impure**. A pure function doesn't rely on external factors and doesn't cause any side effects. Side effects are instead isolated and handled with impure functions. For example:
```
Comment (metagrover) [inline content/blog/functional-programming-constructs/index.md]: (optional)

```suggestion
One of the ways of achieving this is by **favoring pure functions over impure**. A pure function neither relies on any external factors nor causes any side effects. Side effects are instead isolated and handled with impure functions. For example:
```

File: content/blog/functional-programming-constructs/index.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+const addAndMultiply = num => mul2(add1(num));
+
+// use the composed function
+const array = [1, 2, 3].map(num => addAndMultiply(num)); // [4, 6, 8]
+
+// or we can simplify it even more by eliminating the num parameter (make it point-free)
+const anotherArray = [1, 2, 3].map(addAndMultiply); // [4, 6, 8]
+```
+
+The biggest benefit of composition is it leads to better abstractions. Composition also leads to clearer code and reusability as compared to classical inheritance based approach. Here's a quote from the [Functional Light JavaScript](https://github.com/getify/Functional-Light-JS) book, which has become my favorite explanation for abstraction:
```
Comment (metagrover) [inline content/blog/functional-programming-constructs/index.md]: (optional)

```suggestion
The biggest benefit of composition is: it leads to better abstractions. Composition also leads to clearer code and reusability as compared to classical inheritance based approach. Here's a quote from the [Functional Light JavaScript](https://github.com/getify/Functional-Light-JS) book, which has become my favorite explanation for abstraction:
```

**Review by metagrover — APPROVED** [review]
Good stuff!

File: content/blog/functional-programming-constructs/index.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+// Create a new array from an existing array incrementing each value by 1
+// [1, 2, 3] -> [2, 3, 4]
+
+const originalArray = [1, 2, 3];
+
+// First approach using a loop
+let newArray = [];
+// Note: don't use for...in for iterating over arrays since it returns array indices unordered
+for (let value of originalArray) {
+	newArray.push(value + 1);
```
Comment (sagarkarira) [inline content/blog/functional-programming-constructs/index.md:33]: fix indentation

Comment (divyanshu013) [reply content/blog/functional-programming-constructs/index.md:33]: The file runs through prettier, so indentation is auto formatted.

**Pull request merged**